File: frontend/src/api/preferences.ts
import { apiClient } from "@/api/client";

export interface UserPreferences {
  locale: "zh-TW" | "en-US";
  theme: "light" | "dark" | "auto";
  timezone: string;
  calendar: "gregorian" | "minguo";
  page_size: number;
  // 每張表要顯示哪些欄位 (key 由 view 自訂，e.g. "addresses")
  table_columns: Record<string, string[]> | null;
  // last_seen 超過多久就視為離線 (分鐘)；預設 30
  online_grace_minutes: number;
  // Dashboard 常用子網路 UUID 清單
  pinned_subnet_ids: string[] | null;
}

export async function getPreferences(): Promise<UserPreferences> {
  const { data } = await apiClient.get<UserPreferences>("/api/v1/me/preferences");
  return data;
}

export async function updatePreferences(
  patch: Partial<UserPreferences>,
): Promise<UserPreferences> {
  const { data } = await apiClient.patch<UserPreferences>("/api/v1/me/preferences", patch);
  return data;
}

File: backend/app/api/v1/endpoints/preferences.py
"""User preferences endpoints（屬於本人）。"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.v1.dependencies import CurrentUser
from app.core.db import get_session
from app.models.user import UserPreference
from app.schemas.preferences import UserPreferenceRead, UserPreferenceUpdate

router = APIRouter(prefix="/me", tags=["me"])


def _defaults() -> UserPreferenceRead:
    return UserPreferenceRead(
        locale="zh-TW",
        theme="auto",
        timezone="Asia/Taipei",
        calendar="gregorian",
        page_size=50,
        table_columns=None,
        online_grace_minutes=30,
        pinned_subnet_ids=None,
    )


@router.get("/preferences", response_model=UserPreferenceRead)
async def get_preferences(
    user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> UserPreferenceRead:
    pref = await session.get(UserPreference, user.id)
    if pref is None:
        return _defaults()
    return UserPreferenceRead.model_validate(pref)


@router.patch("/preferences", response_model=UserPreferenceRead)
async def update_preferences(
    payload: UserPreferenceUpdate,
    user: CurrentUser,
    session: Annotated[AsyncSession, Depends(get_session)],
) -> UserPreferenceRead:
    pref = await session.get(UserPreference, user.id)
    # mode="json" → UUID / datetime 等型別 dump 成 JSON-safe 字串；
    # pinned_subnet_ids 是 JSONB column，需要 list[str] 才存得進去
    changes = payload.model_dump(exclude_unset=True, mode="json")

    if pref is None:
        defaults = _defaults().model_dump()
        defaults.update(changes)
        pref = UserPreference(user_id=user.id, **defaults)
        session.add(pref)
    else:
        for k, v in changes.items():
            setattr(pref, k, v)

    await session.commit()
    await session.refresh(pref)
    return UserPreferenceRead.model_validate(pref)
